Unable to find contextual data of type: org.keycloak.models.KeycloakSession

### Before reporting an issue

- [x] I have read and understood the above terms for submitting issues, and I understand that my issue may be closed without action if I do not follow them.

### Area

testsuite

### Describe the bug

Most test in oid4vci fail in some after test logic (i.e. AuthServerTestEnricher.removeTestRealms)

```
08:59:02,594 INFO  [org.keycloak.testsuite.arquillian.AuthServerTestEnricher] removing test realms after test class
08:59:02,675 ERROR [io.undertow.request] UT005023: Exception handling request to /auth/admin/realms/test
org.jboss.resteasy.spi.LoggableFailure: RESTEASY003880: Unable to find contextual data of type: org.keycloak.models.KeycloakSession
	at org.jboss.resteasy.core.ContextParameterInjector$GenericDelegatingProxy.invoke(ContextParameterInjector.java:137)
	at jdk.proxy2/jdk.proxy2.$Proxy139.getKeycloakSessionFactory(Unknown Source)
	at org.keycloak.storage.StoreSyncEvent.fire(StoreSyncEvent.java:47)
	at org.keycloak.services.managers.RealmManager$1.commitImpl(RealmManager.java:310)
	at org.keycloak.models.AbstractKeycloakTransaction.commit(AbstractKeycloakTransaction.java:46)
	at org.keycloak.services.DefaultKeycloakTransactionManager.lambda$commitWithTracing$1(DefaultKeycloakTransactionManager.java:171)
	at org.keycloak.tracing.NoopTracingProvider.trace(NoopTracingProvider.java:59)
	at org.keycloak.tracing.NoopTracingProvider.trace(NoopTracingProvider.java:69)
	at org.keycloak.services.DefaultKeycloakTransactionManager.commitWithTracing(DefaultKeycloakTransactionManager.java:170)
	at org.keycloak.services.DefaultKeycloakTransactionManager.lambda$commit$0(DefaultKeycloakTransactionManager.java:147)
	at org.keycloak.tracing.NoopTracingProvider.trace(NoopTracingProvider.java:59)
	at org.keycloak.tracing.NoopTracingProvider.trace(NoopTracingProvider.java:69)
	at org.keycloak.services.DefaultKeycloakTransactionManager.commit(DefaultKeycloakTransactionManager.java:122)
	at org.keycloak.services.DefaultKeycloakSession.closeTransactionManager(DefaultKeycloakSession.java:393)
	at org.keycloak.services.DefaultKeycloakSession.close(DefaultKeycloakSession.java:358)
	at org.keycloak.models.utils.KeycloakModelUtils.runJobInTransactionWithResult(KeycloakModelUtils.java:470)
	at org.keycloak.models.utils.KeycloakModelUtils.runJobInTransaction(KeycloakModelUtils.java:348)
	at org.keycloak.models.utils.KeycloakModelUtils.runJobInTransaction(KeycloakModelUtils.java:338)
```

and

```
[ERROR] org.keycloak.testsuite.oid4vc.issuance.signing.OID4VCTimeNormalizationTest -- Time elapsed: 0.270 s <<< ERROR!
jakarta.ws.rs.InternalServerErrorException: HTTP 500 Internal Server Error
	at org.jboss.resteasy.client.jaxrs.internal.ClientInvocation.handleErrorStatus(ClientInvocation.java:250)
	at org.jboss.resteasy.client.jaxrs.internal.proxy.extractors.DefaultEntityExtractorFactory$3.extractEntity(DefaultEntityExtractorFactory.java:41)
	at org.jboss.resteasy.client.jaxrs.internal.proxy.ClientInvoker.invokeSync(ClientInvoker.java:136)
	at org.jboss.resteasy.client.jaxrs.internal.proxy.ClientInvoker.invoke(ClientInvoker.java:103)
	at org.jboss.resteasy.client.jaxrs.internal.proxy.ClientProxy.invoke(ClientProxy.java:102)
	at jdk.proxy2/jdk.proxy2.$Proxy197.remove(Unknown Source)
	at org.keycloak.testsuite.arquillian.AuthServerTestEnricher.removeTestRealms(AuthServerTestEnricher.java:727)
	at org.keycloak.testsuite.arquillian.AuthServerTestEnricher.afterClass(AuthServerTestEnricher.java:691)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at org.jboss.arquillian.core.impl.ObserverImpl.invoke(ObserverImpl.java:86)
	at org.jboss.arquillian.core.impl.EventContextImpl.invokeObservers(EventContextImpl.java:103)
	at org.jboss.arquillian.core.impl.EventContextImpl.proceed(EventContextImpl.java:90)
	at org.jboss.arquillian.test.impl.TestContextHandler.createClassContext(TestContextHandler.java:83)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at org.jboss.arquillian.core.impl.ObserverImpl.invoke(ObserverImpl.java:86)
	at org.jboss.arquillian.core.impl.EventContextImpl.proceed(EventContextImpl.java:95)
	at org.jboss.arquillian.test.impl.TestContextHandler.createSuiteContext(TestContextHandler.java:69)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at org.jboss.arquillian.core.impl.ObserverImpl.invoke(ObserverImpl.java:86)
	at org.jboss.arquillian.core.impl.EventContextImpl.proceed(EventContextImpl.java:95)
	at org.jboss.arquillian.core.impl.ManagerImpl.fire(ManagerImpl.java:134)
	at org.jboss.arquillian.core.impl.ManagerImpl.fire(ManagerImpl.java:106)
	at org.jboss.arquillian.test.impl.EventTestRunnerAdaptor.afterClass(EventTestRunnerAdaptor.java:95)
	at org.jboss.arquillian.junit.Arquillian$3$1.evaluate(Arquillian.java:177)
	at org.jboss.arquillian.junit.Arquillian.multiExecute(Arquillian.java:344)
	at org.jboss.arquillian.junit.Arquillian.access$200(Arquillian.java:48)
	at org.jboss.arquillian.junit.Arquillian$3.evaluate(Arquillian.java:171)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:54)
```

This happens with current HEAD and still used to work with 7167262909 (Add PK creation for databasechangelog in MySQL to keycloak-database-update.sql when manual migration is used.)

### Version

HEAD

### Regression

- [x] The issue is a regression

### Expected behavior

After test cleanup should work - test realm should be removed

### Actual behavior

Many test fail because of the above. Not sure why this isn't happening in GitHub Actions

### How to Reproduce?

```
cd testsuite/integration-arquillian/tests/base && mvn test -Dtest='org.keycloak.testsuite.oid4vc.**'
```

### Anything else?

_No response_

--- a/model/storage-private/src/main/java/org/keycloak/storage/UserStorageEventListener.java
+++ b/model/storage-private/src/main/java/org/keycloak/storage/UserStorageEventListener.java
@@ -69,12 +69,14 @@ public final class UserStorageEventListener implements ClusterListener, Provider
             });
         } else if (event instanceof StoreSyncEvent ev) {
             UserStorageProviderModel model = ev.getModel() == null ? null: new UserStorageProviderModel(ev.getModel());
-            KeycloakSession session = ev.getSession();
             boolean removed = ev.getRemoved();
-            RealmModel contextRealm = session.getContext().getRealm();
-            RealmModel realm = ev.getRealm();
+            String realmId = ev.getRealm().getId();
 
-            try {
+            runJobInTransaction(sessionFactory, session -> {
+                RealmModel realm = session.realms().getRealm(realmId);
+                if (realm == null) {
+                    return;
+                }
                 session.getContext().setRealm(realm);
 
                 if (model != null) {
@@ -86,9 +88,7 @@ public final class UserStorageEventListener implements ClusterListener, Provider
                         notifyStoreSyncClusterUpdate(session, realm, fedProvider, removed);
                     });
                 }
-            } finally {
-                session.getContext().setRealm(contextRealm);
-            }
+            });
         }
     }
 

--- a/server-spi-private/src/main/java/org/keycloak/storage/StoreSyncEvent.java
+++ b/server-spi-private/src/main/java/org/keycloak/storage/StoreSyncEvent.java
@@ -32,11 +32,11 @@ public class StoreSyncEvent implements ProviderEvent {
     private final boolean removed;
     private final ComponentModel model;
 
-    private StoreSyncEvent(KeycloakSession session, RealmModel realm, boolean removed) {
+    public StoreSyncEvent(KeycloakSession session, RealmModel realm, boolean removed) {
         this(session, realm, null, removed);
     }
 
-    private StoreSyncEvent(KeycloakSession session, RealmModel realm, ComponentModel model, boolean removed) {
+    public StoreSyncEvent(KeycloakSession session, RealmModel realm, ComponentModel model, boolean removed) {
         this.session = session;
         this.realm = realm;
         this.model = model;

--- a/services/src/main/java/org/keycloak/services/managers/RealmManager.java
+++ b/services/src/main/java/org/keycloak/services/managers/RealmManager.java
@@ -42,6 +42,7 @@ import org.keycloak.models.ClientModel;
 import org.keycloak.models.Constants;
 import org.keycloak.models.ImpersonationConstants;
 import org.keycloak.models.KeycloakSession;
+import org.keycloak.models.KeycloakSessionFactory;
 import org.keycloak.models.ModelDuplicateException;
 import org.keycloak.models.ModelException;
 import org.keycloak.models.OTPPolicy;
@@ -310,11 +311,12 @@ public class RealmManager {
                 authSessions.onRealmRemoved(realm);
             }
 
+            KeycloakSessionFactory sessionFactory = session.getKeycloakSessionFactory();
             session.getTransactionManager().enlistAfterCompletion(new AbstractKeycloakTransaction() {
                 @Override
                 protected void commitImpl() {
                     // Refresh periodic sync tasks for configured storageProviders
-                    StoreSyncEvent.fire(session, realm, true);
+                    sessionFactory.publish(new StoreSyncEvent(session, realm, true));
                 }
 
                 @Override
@@ -683,12 +685,13 @@ public class RealmManager {
                 KeycloakModelUtils.setupDeleteAccount(realm.getClientByClientId(Constants.ACCOUNT_MANAGEMENT_CLIENT_ID));
             }
 
+            KeycloakSessionFactory sessionFactory = session.getKeycloakSessionFactory();
             // enlistAfterCompletion(..) as we need to ensure that the realm is committed to the database before we can update the sync tasks
             session.getTransactionManager().enlistAfterCompletion(new AbstractKeycloakTransaction() {
                 @Override
                 protected void commitImpl() {
                     // Refresh periodic sync tasks for configured storageProviders
-                    StoreSyncEvent.fire(session, realm, false);
+                    sessionFactory.publish(new StoreSyncEvent(session, realm, false));
                 }
 
                 @Override

